File: python/cryptology/classical/transposition/simple/rail_fence.py
"""
Rail Fence Cipher implementation.

The Rail Fence cipher is a transposition cipher that writes the message
in a zigzag pattern along multiple "rails" (rows), then reads it off
in a linear fashion.

The pattern looks like a fence when viewed from the side, hence the name.

Example with 3 rails:
  H . . . O . . . R . .
  . E . L . W . O . L .
  . . L . . . D . . . .

Encryption: Write diagonally, read horizontally
Decryption: Write horizontally, read diagonally

Args:
    text: Plaintext or ciphertext
    rails: Number of rails (rows in the zigzag pattern)
    encrypt: True for encryption, False for decryption
"""


def encrypt(plaintext: str, rails: int) -> str:
    """
    Encrypt plaintext using Rail Fence cipher.
    
    Args:
        plaintext: Text to encrypt (case-insensitive, removes non-alphabetic)
        rails: Number of rails (rows)
    
    Returns:
        Encrypted ciphertext
    
    Example:
        >>> encrypt("HELLO WORLD", 3)
        'HLRLELODLLOW'
        >>> encrypt("ATTACK AT DAWN", 4)
        'AKTDCAATTW AN'
    """
    if rails <= 0:
        raise ValueError("Number of rails must be positive")
    
    if rails == 1:
        # Single rail = no encryption
        return ''.join([c.lower() for c in plaintext if c.isalpha()])
    
    # Clean and normalize text to lowercase
    text = ''.join([c.lower() for c in plaintext if c.isalpha()])
    
    if not text:
        return ""
    
    # Build zigzag pattern
    # Each rail stores characters at positions following the zigzag
    pattern = [[] for _ in range(rails)]
    direction = 1  # 1 for down, -1 for up
    rail = 0
    
    for char in text:
        pattern[rail].append(char)
        
        # Change direction at the edges
        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1
        
        rail += direction
    
    # Concatenate all rails
    result = ''.join([''.join(rail) for rail in pattern])
    
    return result


def decrypt(ciphertext: str, rails: int) -> str:
    """
    Decrypt ciphertext using Rail Fence cipher.
    
    Args:
        ciphertext: Encrypted text
        rails: Number of rails (rows)
    
    Returns:
        Decrypted plaintext
    
    Example:
        >>> decrypt("HLRLELODLLOW", 3)
        'HELLOWORLD'
    """
    if rails <= 0:
        raise ValueError("Number of rails must be positive")
    
    if rails == 1:
        return ''.join([c.lower() for c in ciphertext if c.isalpha()])
    
    # Clean the ciphertext to lowercase
    text = ''.join([c.lower() for c in ciphertext if c.isalpha()])
    
    if not text:
        return ""
    
    # Calculate how many characters are in each rail
    # Simulate the pattern to count characters per rail
    pattern_lengths = [0] * rails
    direction = 1
    rail = 0
    
    for i in range(len(text)):
        pattern_lengths[rail] += 1
        
        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1
        
        rail += direction
    
    # Distribute ciphertext across rails
    pattern = [[] for _ in range(rails)]
    text_pos = 0
    
    for rail_num in range(rails):
        for _ in range(pattern_lengths[rail_num]):
            if text_pos < len(text):
                pattern[rail_num].append(text[text_pos])
                text_pos += 1
    
    # Read off in zigzag order
    result = []
    direction = 1
    rail = 0
    rail_positions = [0] * rails
    
    for _ in range(len(text)):
        if rail_positions[rail] < len(pattern[rail]):
            result.append(pattern[rail][rail_positions[rail]])
            rail_positions[rail] += 1
        
        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1
        
        rail += direction
    
    return ''.join(result)


def _visualize_pattern(text: str, rails: int) -> str:
    """
    Visualize the rail fence pattern (for debugging).
    
    Args:
        text: Text to visualize
        rails: Number of rails
    
    Returns:
        String representation of the pattern
    """
    # Clean text to lowercase
    clean_text = ''.join([c.lower() for c in text if c.isalpha()])
    
    # Create grid
    rows = len(clean_text)
    grid = [[' ' for _ in range(rows)] for _ in range(rails)]
    
    direction = 1
    rail = 0
    
    for col, char in enumerate(clean_text):
        grid[rail][col] = char
        
        if rail == 0:
            direction = 1
        elif rail == rails - 1:
            direction = -1
        
        rail += direction
    
    # Convert to string
    result = []
    for row in grid:
        result.append(' '.join(row))
    
    return '\n'.join(result)


def get_key_range(text_length: int) -> tuple:
    """
    Get the valid range of keys for a given text length.
    
    Args:
        text_length: Length of the text
    
    Returns:
        Tuple of (min_key, max_key)
    """
    if text_length <= 0:
        return (0, 0)
    return (1, min(text_length, 10))  # Practical limit


if __name__ == "__main__":
    # Test the implementation
    print("=" * 60)
    print("RAIL FENCE CIPHER")
    print("=" * 60)
    
    plaintext = "HELLO WORLD"
    rails = 3
    
    print(f"\nPlaintext:  {plaintext}")
    print(f"Rails:      {rails}")
    print(f"Plaintext (cleaned): {''.join([c.lower() for c in plaintext if c.isalpha()])}")
    
    encrypted = encrypt(plaintext, rails)
    decrypted = decrypt(encrypted, rails)
    
    print(f"\nEncrypted:  {encrypted}")
    print(f"Decrypted:  {decrypted}")
    print(f"Success:    {decrypted == ''.join([c.lower() for c in plaintext if c.isalpha()])}")
    
    print("\nVisualization:")
    print(_visualize_pattern(plaintext, rails))
    
    # Test with different number of rails
    print("\n" + "=" * 60)
    plaintext2 = "ATTACK AT DAWN"
    rails2 = 4
    
    print(f"\nPlaintext:  {plaintext2}")
    print(f"Rails:      {rails2}")
    
    encrypted2 = encrypt(plaintext2, rails2)
    decrypted2 = decrypt(encrypted2, rails2)
    
    print(f"Encrypted:  {encrypted2}")
    print(f"Decrypted:  {decrypted2}")
    print(f"Success:    {decrypted2 == ''.join([c.lower() for c in plaintext2 if c.isalpha()])}")

File: c/src/classical/transposition/simple/rail_fence.c
#include "cryptology/classical/transposition/simple/rail_fence.h"
#include "cryptology/classical/transposition/utf8_helpers.h"
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

int rail_fence_encrypt(const char *plaintext, int rails, char *output, size_t output_size) {
    if (!plaintext || !output || rails <= 0 || output_size == 0) {
        return -1;
    }

    // Clean and normalize text (UTF-8 aware, supports Turkish)
    char *clean_input = malloc(strlen(plaintext) * 3 + 1);
    if (!clean_input) return -1;

    if (clean_utf8_text(plaintext, clean_input, strlen(plaintext) * 3 + 1) != 0) {
        free(clean_input);
        return -1;
    }

    if (strlen(clean_input) == 0) {
        output[0] = '\0';
        free(clean_input);
        return 0;
    }

    if (rails == 1) {
        // Single rail = no encryption
        strncpy(output, clean_input, output_size - 1);
        output[output_size - 1] = '\0';
        free(clean_input);
        return 0;
    }

    size_t text_len = strlen(clean_input);

    // Build zigzag pattern
    char **pattern = malloc(rails * sizeof(char*));
    if (!pattern) {
        free(clean_input);
        return -1;
    }

    for (int i = 0; i < rails; i++) {
        pattern[i] = malloc(text_len * sizeof(char));
        if (!pattern[i]) {
            // Clean up allocated memory
            for (int j = 0; j < i; j++) {
                free(pattern[j]);
            }
            free(pattern);
            free(clean_input);
            return -1;
        }
        memset(pattern[i], 0, text_len);
    }

    int direction = 1;  // 1 for down, -1 for up
    int rail = 0;

    for (size_t i = 0; i < text_len; i++) {
        pattern[rail][i] = clean_input[i];

        // Change direction at the edges
        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }

        rail += direction;
    }

    // Concatenate all rails
    size_t output_index = 0;
    for (int rail_num = 0; rail_num < rails && output_index < output_size - 1; rail_num++) {
        for (size_t i = 0; i < text_len && output_index < output_size - 1; i++) {
            if (pattern[rail_num][i] != 0) {
                output[output_index++] = pattern[rail_num][i];
            }
        }
    }
    output[output_index] = '\0';

    // Clean up
    for (int i = 0; i < rails; i++) {
        free(pattern[i]);
    }
    free(pattern);
    free(clean_input);

    return 0;
}

int rail_fence_decrypt(const char *ciphertext, int rails, char *output, size_t output_size) {
    if (!ciphertext || !output || rails <= 0 || output_size == 0) {
        return -1;
    }

    // Clean the ciphertext (UTF-8 aware, supports Turkish)
    char *clean_input = malloc(strlen(ciphertext) * 3 + 1);
    if (!clean_input) return -1;

    if (clean_utf8_text(ciphertext, clean_input, strlen(ciphertext) * 3 + 1) != 0) {
        free(clean_input);
        return -1;
    }

    if (strlen(clean_input) == 0) {
        output[0] = '\0';
        free(clean_input);
        return 0;
    }

    if (rails == 1) {
        strncpy(output, clean_input, output_size - 1);
        output[output_size - 1] = '\0';
        free(clean_input);
        return 0;
    }

    size_t text_len = strlen(clean_input);

    // Calculate how many characters are in each rail
    int *pattern_lengths = malloc(rails * sizeof(int));
    if (!pattern_lengths) {
        free(clean_input);
        return -1;
    }
    memset(pattern_lengths, 0, rails * sizeof(int));

    int direction = 1;
    int rail = 0;

    for (size_t i = 0; i < text_len; i++) {
        pattern_lengths[rail]++;

        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }

        rail += direction;
    }

    // Distribute ciphertext across rails
    char **pattern = malloc(rails * sizeof(char*));
    if (!pattern) {
        free(pattern_lengths);
        free(clean_input);
        return -1;
    }

    for (int i = 0; i < rails; i++) {
        pattern[i] = malloc(pattern_lengths[i] * sizeof(char));
        if (!pattern[i]) {
            // Clean up allocated memory
            for (int j = 0; j < i; j++) {
                free(pattern[j]);
            }
            free(pattern);
            free(pattern_lengths);
            free(clean_input);
            return -1;
        }
    }

    size_t text_pos = 0;
    for (int rail_num = 0; rail_num < rails; rail_num++) {
        for (int i = 0; i < pattern_lengths[rail_num]; i++) {
            if (text_pos < text_len) {
                pattern[rail_num][i] = clean_input[text_pos++];
            }
        }
    }

    // Read off in zigzag order
    size_t output_index = 0;
    direction = 1;
    rail = 0;
    int *rail_positions = malloc(rails * sizeof(int));
    if (!rail_positions) {
        // Clean up
        for (int i = 0; i < rails; i++) {
            free(pattern[i]);
        }
        free(pattern);
        free(pattern_lengths);
        free(clean_input);
        return -1;
    }
    memset(rail_positions, 0, rails * sizeof(int));

    for (size_t i = 0; i < text_len && output_index < output_size - 1; i++) {
        if (rail_positions[rail] < pattern_lengths[rail]) {
            output[output_index++] = pattern[rail][rail_positions[rail]];
            rail_positions[rail]++;
        }

        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }

        rail += direction;
    }
    output[output_index] = '\0';

    // Clean up
    for (int i = 0; i < rails; i++) {
        free(pattern[i]);
    }
    free(pattern);
    free(pattern_lengths);
    free(rail_positions);
    free(clean_input);

    return 0;
}

int rail_fence_visualize(const char *text, int rails, char *output, size_t output_size) {
    if (!text || !output || rails <= 0 || output_size == 0) {
        return -1;
    }

    // Clean text (UTF-8 aware, supports Turkish)
    char *clean_input = malloc(strlen(text) * 3 + 1);
    if (!clean_input) return -1;

    if (clean_utf8_text(text, clean_input, strlen(text) * 3 + 1) != 0) {
        free(clean_input);
        return -1;
    }

    size_t text_len = strlen(clean_input);
    if (text_len == 0) {
        output[0] = '\0';
        free(clean_input);
        return 0;
    }

    // Create grid
    char **grid = malloc(rails * sizeof(char*));
    if (!grid) {
        free(clean_input);
        return -1;
    }

    for (int i = 0; i < rails; i++) {
        grid[i] = malloc(text_len * sizeof(char));
        if (!grid[i]) {
            // Clean up allocated memory
            for (int j = 0; j < i; j++) {
                free(grid[j]);
            }
            free(grid);
            free(clean_input);
            return -1;
        }
        memset(grid[i], ' ', text_len);
    }

    int direction = 1;
    int rail = 0;

    for (size_t col = 0; col < text_len; col++) {
        grid[rail][col] = clean_input[col];

        if (rail == 0) {
            direction = 1;
        } else if (rail == rails - 1) {
            direction = -1;
        }

        rail += direction;
    }

    // Convert to string
    size_t output_index = 0;
    for (int row = 0; row < rails && output_index < output_size - 1; row++) {
        for (size_t col = 0; col < text_len && output_index < output_size - 1; col++) {
            output[output_index++] = grid[row][col];
            if (col < text_len - 1 && output_index < output_size - 1) {
                output[output_index++] = ' ';
            }
        }
        if (row < rails - 1 && output_index < output_size - 1) {
            output[output_index++] = '\n';
        }
    }
    output[output_index] = '\0';

    // Clean up
    for (int i = 0; i < rails; i++) {
        free(grid[i]);
    }
    free(grid);
    free(clean_input);

    return 0;
}

void rail_fence_get_key_range(size_t text_length, int *min_key, int *max_key) {
    if (text_length <= 0) {
        *min_key = 0;
        *max_key = 0;
    } else {
        *min_key = 1;
        *max_key = (int)(text_length < 10 ? text_length : 10);  // Practical limit
    }
}
